packages/sdk-js/src/plugins/browser-performance.ts

```diff
@@ -0,0 +1,189 @@
+import type { GrowthBook } from "../GrowthBook";
+import type { UserScopedGrowthBook } from "../GrowthBookClient";
+
+type BrowserPerformanceSettings = {
+  trackFCP?: boolean;
+  trackLCP?: boolean;
+  trackFID?: boolean;
+  trackCLS?: boolean;
+  trackTTFB?: boolean;
+  trackTBT?: boolean;
+  trackErrors?: boolean;
+};
+
+export function browserPerformancePlugin({
+  trackFCP = true,
+  trackLCP = true,
+  trackFID = true,
+  trackCLS = true,
+  trackTTFB = true,
+  trackTBT = true,
+  trackErrors = true,
+}: BrowserPerformanceSettings = {}) {
+  if (typeof window === "undefined" || typeof document === "undefined") {
+    throw new Error("browserPerformancePlugin only works in the browser");
+  }
+
+  return (gb: GrowthBook | UserScopedGrowthBook) => {
+    if (!gb.logEvent) {
+      throw new Error("GrowthBook instance must have a logEvent method");
+    }
+
+    if ("PerformanceObserver" in window) {
+      try {
+        let observing = true;
+        const observers: PerformanceObserver[] = [];
+        const stopObserving = () => {
+          observing = false;
+          observers.forEach((observer) => observer.disconnect());
+        };
+
+        let lcpTime = 0;
+        let clsValue = 0;
+        let tbtValue = 0;
+
+        const reportCWV = () => {
+          if (!observing) return;
+          stopObserving();
+          if (trackLCP && lcpTime) {
+            gb.logEvent("CWV:LCP", { value: lcpTime });
+          }
+          if (trackCLS && clsValue) {
+            gb.logEvent("CWV:CLS", { value: clsValue });
+          }
+          if (trackTBT && tbtValue) {
+            gb.logEvent("CWV:TBT", { value: tbtValue });
+          }
+          if ("flushEventLog" in gb) {
+            // @ts-expect-error instantly flush the tracking event log, otherwise there is a race condition with flush()
+            gb?.flushEventLog?.();
```

Comment:
We defer long-running metrics until visibility change / close. This is a race condition with `growthbook-tracking` which also listens to visibility change to flush the event queue. We overcome this by writing the `flush()` handler onto the SDK directly so that we can call it instantly. Not pretty, but not sure there's a better way, since these plugins don't expose an API.

Reply:
removed this hook and instead changed the tracking lib to auto-flush the queue once it has started "unloading"